dart/Colfer.dart

```diff
@@ -0,0 +1,1226 @@
+  /// A tests binaries.
+  Uint8List a;
+
+  /// O tests nested data structures.
+  O? o;
+
+  /// Os tests data structure lists.
+  List<O?> os;
+
+  /// Ss tests text lists.
+  List<String> ss;
+
+  /// As tests binary lists.
+  List<Uint8List> as_;
+
+  /// U8 tests unsigned 8-bit integers.
+  int u8;
+
+  /// U16 tests unsigned 16-bit integers.
+  int u16;
+
+  /// F32s tests 32-bit floating point lists.
+  Float32List f32s;
+
+  /// F64s tests 64-bit floating point lists.
+  Float64List f64s;
+
+  @override
+  bool operator ==(_other) {
+    return (_other is O) &&
+        _other.b == b &&
+        _other.u32 == u32 &&
+        _other.u64 == u64 &&
+        _other.i32 == i32 &&
+        _other.i64 == i64 &&
+        _other.f32 == f32 &&
+        _other.f64 == f64 &&
+        _other.t == t &&
+        _other.s == s &&
+        IterableEquality().equals(_other.a, a) &&
+        _other.o == o &&
+        IterableEquality().equals(_other.os, os) &&
+        IterableEquality().equals(_other.ss, ss) &&
+        DeepCollectionEquality().equals(_other.as_, as_) &&
+        _other.u8 == u8 &&
+        _other.u16 == u16 &&
+        IterableEquality().equals(_other.f32s, f32s) &&
+        IterableEquality().equals(_other.f64s, f64s);
+  }
+
+  String toString() {
+    return 'class O {'
+            'b: ${b.toString()}'
+            ', u32: ${u32.toString()}'
+            ', u64: ${u64.toString()}'
+            ', i32: ${i32.toString()}'
+            ', i64: ${i64.toString()}'
+            ', f32: ${f32.toString()}'
+            ', f64: ${f64.toString()}'
+            ', t: ${t.toString()}'
+            ', s: "$s"'
+            ', a: ${a.toString()}'
+            ', o: ${o.toString()}'
+            ', os: List<O>${os.toString()}'
+            ', ss: [' +
+        (ss.isNotEmpty ? '"${ss.join('", "')}"' : '') +
+        ']'
+            ', as_: List<Uint8List>${as_.toString()}'
+            ', u8: ${u8.toString()}'
+            ', u16: ${u16.toString()}'
+            ', f32s: List<Float32List>${f32s.toString()}'
+            ', f64s: List<Float64List>${f64s.toString()}';
+  }
+
+  @override
+  int get hashCode =>
+      hashObjects([b, u32, u64, i32, i64, f32, f64, t, s, a, o, os, ss, as_, u8, u16, f32s, f64s]);
+
+  O({
+    this.b = false,
+    this.u32 = 0,
+    this.u64 = 0,
+    this.i32 = 0,
+    this.i64 = 0,
+    this.f32 = 0.0,
+    this.f64 = 0.0,
+    this.t,
+    this.s = '',
+    Uint8List? a,
+    this.o,
+    List<O>? os,
+    List<String>? ss,
+    List<Uint8List>? as_,
+    this.u8 = 0,
+    this.u16 = 0,
+    Float32List? f32s,
+    Float64List? f64s,
+  })  : a = a ?? Uint8List(0),
+        os = os ?? [],
+        ss = ss ?? [],
+        as_ = as_ ?? [],
+        f32s = f32s ?? Float32List(0),
+        f64s = f64s ?? Float64List(0);
+
+  /// Returns an over estimatation of marshal length.
+
+  /// Throws [Exception] if the size of a list exceeds [colferListMax],
+  /// or if a text, binary, or the estimation exceeds [colferSizeMax].
```

Comment:
The `Exception` may be unjust if the UTF-8 guess causes a `colferSizeMax` breach.

https://github.com/pascaldekloe/colfer/blob/aa76077a0fe822cd223b112f547e3319626b3f3b/java/gen/O.java#L251

The marshal method may never produce anything beyond `colferSizeMax`, so that is where the error must be raised.

Reply:
Be careful with overflows here. The Java example doesn't have this issue because of its internal 31-bit limitations.

Reply:
You mean the estimated length `_l` or the exact size `_i` overflowing signed 64 bit integer positive max = `2^63 - 1`? I think it's quite big not to reach by anyone.

Reply:
A 63-bit counter can be assumed as plenty indeed. 32 Bits or less are not enough though. 😅